### PR TITLE
Master purchase whe

### PR DESCRIPTION
**Pull request by pimodoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase_requisition/models/purchase_requisition.py [hunk 1]
```diff
@@ -66,6 +66,25 @@ def _get_type_id(self):
...
+    is_quantity_copy = fields.Boolean(compute='_is_quantity_copy')
+
+    @api.model
+    def create(self, vals):
+        rec = super(PurchaseRequisition, self).create(vals)
+        if rec.name == 'New':
+            if rec.is_quantity_copy:
+                rec.name = self.env['ir.sequence'].next_by_code('purchase.requisition.purchase.tender')
+            else:
+                rec.name = self.env['ir.sequence'].next_by_code('purchase.requisition.blanket.order')
```
Comment (amoyaux) [inline addons/purchase_requisition/models/purchase_requisition.py]: Should be moves into next commit (in order to be atomic)

File: addons/purchase_requisition/models/purchase_requisition.py [hunk 2]
```diff
@@ -150,11 +175,50 @@ class PurchaseRequisitionLine(models.Model):
...
+    supplier_info_ids = fields.One2many('product.supplierinfo', 'purchase_requisition_line_id')
+
+    @api.multi
+    def create(self,vals):
+        res = super(PurchaseRequisitionLine, self).create(vals)
+        if res.requisition_id.state not in ['draft', 'cancel', 'done'] and not res.requisition_id.is_quantity_copy:
+            supplier_infos = self.env['product.supplierinfo'].search([
+                ('product_id', '=', vals.get('product_id')),
+                ('name', '=', res.requisition_id.vendor_id.id),
+            ])
```
Comment (amoyaux) [inline addons/purchase_requisition/models/purchase_requisition.py]: Strange to me. Why don't directly search for the requisition id there? It won't work for 2 purchases requisitions that have the same vendor/same product and won't raise a warning.

File: addons/purchase/models/purchase.py [hunk 3]
```diff
@@ -822,7 +822,7 @@ def _onchange_quantity(self):
         if not self.product_id:
             return
 
-        seller = self.product_id._select_seller(
+        seller = self.product_id.with_context(current_po=self.order_id)._select_seller(
```
Comment (amoyaux) [inline addons/purchase/models/purchase.py]: Gonna replace it by a kwargs tomorrow

Comment (sle-odoo) [reply addons/purchase/models/purchase.py]: huge lie!

Comment (amoyaux) [reply addons/purchase/models/purchase.py]: not cool 👎

File: addons/purchase_requisition/models/purchase_requisition.py [hunk 4]
```diff
@@ -66,6 +66,25 @@ def _get_type_id(self):
                               copy=False, default='draft')
     account_analytic_id = fields.Many2one('account.analytic.account', 'Analytic Account')
     picking_type_id = fields.Many2one('stock.picking.type', 'Operation Type', required=True, default=_get_picking_in)
+    is_quantity_copy = fields.Boolean(compute='_is_quantity_copy')
```
Comment (pimodoo) [inline addons/purchase_requisition/models/purchase_requisition.py]: commit message: The witch won't be happy

File: addons/purchase_requisition/models/purchase_requisition.py [hunk 5]
```diff
@@ -399,6 +414,16 @@ def _onchange_quantity(self):
         return res
 
 
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    def _prepare_sellers(self):
+        if self.env.context and 'current_po' in self.env.context:
```
Comment (sle-odoo) [inline addons/purchase_requisition/models/purchase_requisition.py]: self.env.context.get

File: addons/purchase_requisition/models/purchase_requisition.py (hunk 4 above)
Comment (mreficent) [reply addons/purchase_requisition/models/purchase_requisition.py]: computes should be named <code>_compute_something</code>

File: addons/product/models/product.py [hunk 6]
```diff
@@ -464,18 +464,18 @@ def open_product_template(self):
...
-    def _select_seller(self, partner_id=False, quantity=0.0, date=None, uom_id=False):
+    def _select_seller(self, partner_id=False, quantity=0.0, date=None, uom_id=False, current_po=None):
         self.ensure_one()
         if date is None:
             date = fields.Date.context_today(self)
         precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
 
         res = self.env['product.supplierinfo']
-        for seller in self._prepare_sellers():
+        for seller in self._prepare_sellers(current_po):
```
Comment (sle-odoo) [inline addons/product/models/product.py]: current_po=current_po

File: addons/purchase_requisition/models/purchase_requisition.py [hunk 7]
```diff
@@ -398,6 +414,16 @@ def _onchange_quantity(self):
...
+            return [s for s in self.seller_ids if not s.purchase_requisition_id or s.purchase_requisition_id == self.env.context.get('current_po').requisition_id]
```
Comment (sle-odoo) [inline addons/purchase_requisition/models/purchase_requisition.py]: when creating a PO from a BO never chose the seller of an another BO

File: addons/purchase/models/purchase.py [hunk 8]
```diff
@@ -821,12 +821,13 @@ def onchange_product_id_warning(self):
     def _onchange_quantity(self):
         if not self.product_id:
             return
-
-        seller = self.product_id.with_context(current_po=self.order_id)._select_seller(
+        seller_ids = self.product_id.seller_ids.filtered(lambda s: not s.requisition_id or s.requisition_id.id == self.order_id.id)
```
Comment (sle-odoo) [inline addons/purchase/models/purchase.py]: ref to requisition in purchase

**Pull request merged**